privacyidea/lib/token.py

```diff
@@ -1071,34 +1071,38 @@ def init_token(param, user=None, tokenrealms=None,
     # and to the user realm
     if user and user.realm:
         realms.append(user.realm)
-    if realms or user:
-        # We need to save the token to the DB, otherwise the Token
-        # has no id!
-        db_token.save()
-        db_token.set_realms(realms)
 
-    # the tokenclass object is created
-    tokenobject = create_tokenclass_object(db_token)
+    try:
+        # Save the token to the database
+        if token_count == 0:
+            db_token.save()
 
-    if token_count == 0:
-        # if this token is a newly created one, we have to setup the defaults,
-        # which later might be overwritten by the tokenobject.update(param)
-        tokenobject.set_defaults()
+        # the tokenclass object is created
+        tokenobject = create_tokenclass_object(db_token)
 
-    # Set the user of the token
-    if user is not None and user.login != "":
-        tokenobject.add_user(user)
+        if token_count == 0:
+            # if this token is a newly created one, we have to setup the defaults,
+            # which later might be overwritten by the tokenobject.update(param)
+            tokenobject.set_defaults()
 
-    tokenobject.update(param)
+        # Set the user of the token
+        if user is not None and user.login != "":
+            tokenobject.add_user(user)
 
-    try:
-        # Save the token to the database
-        db_token.save()
+        # Set the token realms (updates the TokenRealm table)
+        if realms or user:
+            db_token.set_realms(realms)
+
+        tokenobject.update(param)
 
     except Exception as e:  # pragma: no cover
-        log.error('token create failed!')
+        log.error('token create failed: {0!s}'.format(e))
         log.debug("{0!s}".format(traceback.format_exc()))
-        raise TokenAdminError(_("token create failed {0!r}").format(e), id=1112)
+        # delete the newly created token from the db
+        if token_count == 0:
+            db_token.delete()
+        raise
```

Comment:
I would reccomend raising the adminError again.

Reply:
The problem is, that before we threw that error only if the commit to the db failed (`db_token.save()`). Now we call several functions in the `try:` block which throw several different errors (`ParameterError`, `TokenAdminError`, ...). Since we don't want to loose there errors we catch them, clean up the database and re-raise them for further processing.